### PR TITLE
SOLR-16104: TestCoreDiscovery.testTooManyTransientCores() to use retryUntil logic

### PR DESCRIPTION
**Pull request by cpoerschke**
https://issues.apache.org/jira/browse/SOLR-16104

### CODE REVIEW

File: solr/core/src/test/org/apache/solr/core/TestCoreDiscovery.java [hunk 1]
```diff
@@ -328,7 +328,7 @@ public void testTooManyTransientCores() throws Exception {
       // Just check that the proper number of cores are loaded since making the test depend on order
       // would be fragile
       assertEquals(
-          "There should only be 3 cores loaded, coreLOS and two coreT? cores",
+          "There should only be 3 cores loaded, coreLOS and two coreT? cores. Actual: "+cc.getLoadedCoreNames(),
```
Comment (cpoerschke) [inline solr/core/src/test/org/apache/solr/core/TestCoreDiscovery.java:331]: The code at https://github.com/apache/solr/blob/0c73f3e64f46e8ae1134f2f19f8f0ad29990c9e5/solr/core/src/java/org/apache/solr/core/TransientSolrCoreCacheDefault.java#L94 suggests that 3 as an absolute limit here is tricky.

Could the test perhaps poll and accept "3 or 4 but eventually 3" values?

**Comment by dsmiley** [comment]
I'm not sure allowing 4 is in the spirit of the test's purpose.  Instead, I think the assertion should be re-attempted after some time.  I tried this locally:
```java
  private void assertEventually(int upToTime, TimeUnit unit, Runnable task) {
    final long startNanos = System.nanoTime();
    for (int attempt = 1; true; attempt++) {
      try {
        task.run();
      } catch (AssertionError | RuntimeException e) {
        final long elapsedNanos = System.nanoTime() - startNanos;
        final long remainingNanos = unit.toNanos(upToTime) - elapsedNanos;
        if (remainingNanos <= 0) {
          throw e;
        }
        try {
          // sleep with exponential back-off.  Don't exceed remaining time. Not shorter than 5 ms.
          final long sleepNanos = Math.min(Math.max(5000, elapsedNanos), remainingNanos);
          Thread.sleep(TimeUnit.NANOSECONDS.toMillis(sleepNanos));
        } catch (InterruptedException ie) {
          throw e; // throw original exception, not the interrupted one
        }
        System.err.println("Retry attempt " + attempt);
        continue; // try again
      }
      return; // success!
    }
  }
```

And debugged it a bit to ensure it does what it's supposed to do.  I think this will work for this scenario and we could even add this utility method somewhere if we want it elsewhere.  The name is debatable; could be retryUntil.  I found `org.apache.solr.common.util.RetryUtil` which is pretty similar.  Come to think of it, maybe we should just call that existing code -- `retryUntil` ?

(I debated to create a whole new PR or push commits reverting your change, or to just comment here so I just put here as a comment for now)

**Comment by dsmiley** [comment]
we don't even need a CHANGES.txt; it's a test fix.  I suggest we merge this and see what happens over the next week or two; ok?

**Comment by cpoerschke** [comment]
> we don't even need a CHANGES.txt; it's a test fix. I suggest we merge this and see what happens over the next week or two; ok?

Sounds good. Thanks!

**Pull request merged**